gson/src/main/java/com/google/gson/JsonPrimitive.java

```diff
@@ -290,11 +290,10 @@ public boolean equals(Object obj) {
           : this.getAsNumber().longValue() == other.getAsNumber().longValue();
     }
     if (value instanceof Number && other.value instanceof Number) {
-      double a = getAsNumber().doubleValue();
-      // Java standard types other than double return true for two NaN. So, need
-      // special handling for double.
-      double b = other.getAsNumber().doubleValue();
-      return a == b || (Double.isNaN(a) && Double.isNaN(b));
+      return this.value instanceof BigDecimal || other.value instanceof BigDecimal
+          ? this.getAsBigDecimal().equals(other.getAsBigDecimal())
```

Comment:
Another problem here might be comparing `NaN`, `NEGATIVE_INFINITY` and `POSITIVE_INFINITY` (for `Double` and `Float`) with a `BigDecimal`. In the past that made `equals` return `false` (except for rounding issues maybe for large `BigDecimal`), while with the code changes of this pull request it would throw an exception.

So maybe if one value is a `BigDecimal` but the other is NaN or Infinity `equals` should directly return `false`?